source/extensions/health_checkers/thrift/thrift.cc

```diff
@@ -0,0 +1,119 @@
+      method_name_(thrift_config.method_name()),
+      transport_(TransportNames::get().getTypeFromProto(thrift_config.transport())),
+      protocol_(ProtocolNames::get().getTypeFromProto(thrift_config.protocol())),
+      client_factory_(client_factory) {
+  if (transport_ == TransportType::Auto || protocol_ == ProtocolType::Auto ||
+      protocol_ == ProtocolType::Twitter) {
+    throw EnvoyException(
+        fmt::format("Invalid thrift health check configuration: {}", thrift_config.DebugString()));
+  }
+}
+
+// ThriftActiveHealthCheckSession:
+ThriftHealthChecker::ThriftActiveHealthCheckSession::ThriftActiveHealthCheckSession(
+    ThriftHealthChecker& parent, const Upstream::HostSharedPtr& host)
+    : ActiveHealthCheckSession(parent, host), parent_(parent),
+      hostname_(getHostname(host, parent_.cluster_.info())) {
+  ENVOY_LOG(trace, "ThriftActiveHealthCheckSession construct hostname={}", hostname_);
+}
+
+ThriftHealthChecker::ThriftActiveHealthCheckSession::~ThriftActiveHealthCheckSession() {
+  ENVOY_LOG(trace, "ThriftActiveHealthCheckSession destruct");
+  ASSERT(client_ == nullptr);
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onDeferredDelete() {
+  if (client_) {
+    expect_close_ = true;
+    client_->close();
+  }
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onInterval() {
+  ENVOY_LOG(trace, "ThriftActiveHealthCheckSession onInterval");
+  if (!client_) {
+    ENVOY_LOG(trace, "ThriftActiveHealthCheckSession construct client");
+    client_ = parent_.client_factory_.create(*this, parent_.transport_, parent_.protocol_,
+                                             parent_.method_name_, host_, parent_.random_.random());
+    client_->start();
+    expect_close_ = false;
+  }
+
+  client_->makeRequest();
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onTimeout() {
+  expect_close_ = true;
+  client_->close();
+}
+
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onResponseResult(bool is_success) {
+  if (is_success) {
+    handleSuccess();
+  } else {
+    // TODO(kuochunghsu): We might want to define retriable response.
+    handleFailure(envoy::data::core::v3::ACTIVE, /* retriable */ false);
+  }
+}
+
+Upstream::Host::CreateConnectionData
+ThriftHealthChecker::ThriftActiveHealthCheckSession::createConnection() {
+  return host_->createHealthCheckConnection(parent_.dispatcher_, parent_.transportSocketOptions(),
+                                            parent_.transportSocketMatchMetadata().get());
+}
+// Network::ConnectionCallbacks
+void ThriftHealthChecker::ThriftActiveHealthCheckSession::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    ENVOY_LOG(trace, "on event close, is_local_close={} expect_close={}",
+              event == Network::ConnectionEvent::LocalClose, expect_close_);
+    // TODO: check if response is partial or complete
```

Comment:
TODO(kuochung)

Reply:
i remove this since time_out check should be enough